lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp

```diff
@@ -88,6 +88,43 @@ LinearLayout identityND(StringAttr inDimName, ArrayRef<unsigned> shape,
   return ret;
 }
 
+// Returns 2D layout that traverses the 2nd leading dimension first and then
+// 1st leading dimension. Only supports 2D layout for now.
+LinearLayout transposeND(StringAttr inDimName, ArrayRef<unsigned> shape,
+                         ArrayRef<unsigned> order,
+                         ArrayRef<StringAttr> outDimNames) {
+  assert(shape.size() == order.size());
+  assert((order.size() == 2) && "only support dim of 2 now");
+
+  MLIRContext *ctx = inDimName.getContext();
+  StringAttr kRegister = S("register");
+  StringAttr kDim0 = S("dim0");
+  StringAttr kDim1 = S("dim1");
+
+  LinearLayout ret = LinearLayout::empty();
+  std::vector<std::vector<int32_t>> bases;
+  // traverse 2nd dimension (K-dim in GEMM case)
+  int dim = order[1];
+  for (int basis = 1; basis < shape[dim]; basis <<= 1) {
+    bases.push_back({0, basis});
+  }
+  // traverse 1st dimension (N-dim in GEMM K-major B-tensor)
+  // this is the consecutive dimension loaded from global memory
+  dim = order[0];
+  for (int basis = 1; basis < shape[dim]; basis <<= 1) {
+    bases.push_back({basis, 0});
+  }
+  auto dimMinor = "dim" + std::to_string(order[0]);
+  auto dimMajor = "dim" + std::to_string(order[1]);
+  StringAttr kDimMinor = S(dimMinor);
+  StringAttr kDimMajor = S(dimMajor);
+  ret = LinearLayout(
+      {{kRegister, bases}},
+      {{kDimMinor, shape[order[0]]}, {kDimMajor, shape[order[1]]}}, false);
```

Comment:
You don't need to "unroll" the loop. Try to do it in a similar way as in identityND